### PR TITLE
improve random generation

### PR DESCRIPTION
**Pull request by imsamurai**
1. generate random value by pattern regexp
2. fix random number for case of negative numbers
3. fix allOf string pattern serialization

### CODE REVIEW

**Comment by Gi60s** [comment]
This PR looks really good. Thanks for sharing it. I've got a few other changes that should be ready in the next few days and I'll roll this PR and those together into one release.

**Review by stuft2 — COMMENTED** [review]
@Gi60s would you take a look at the comments I left and let me know what you think?

File: src/enforcers/Schema.js [hunk 1]
```diff
@@ -1204,6 +1204,9 @@ function numericType (schema) {
 }
 
 function serializeSchema (schema, exception, dataTypes, serializedSchemas) {
+    if (schema.pattern) {
+        schema.pattern = schema.pattern.source;
```
[the remark this answers is not in the document]
Comment (Gi60s) [reply src/enforcers/Schema.js:1208]: I believe that this is what we want to be doing here. This is a method that takes a built schema and attempts to create a definition from it, so this line is converting the regex object back into a string.

File: src/util.js [hunk 2]
```diff
@@ -417,15 +417,20 @@ function randomNumber ({ min, max, multipleOf, exclusiveMin = false, exclusiveMa
     const minIsNumber = isNumber(min);
     const maxIsNumber = isNumber(max);
 
+    if (max < min) {
+        max = 2 * min;
```
[the remark this answers is not in the document]
Comment (Gi60s) [reply src/util.js:421]: I agree that it should throw an error.

I found a bug in the schema's run random that could feed an invalid maximum or minimum value into the `util.js` random number function and I fixed that there.

I'm changing this file to throw an error so we can catch if there are other places where invalid max and min values are passed in.

**Comment by Gi60s** [comment]
Changes have been published to NPM as version `1.14.0`. Thanks for the PR.

**Pull request merged**